lib/event_tracer/appsignal_logger.rb

```diff
@@ -2,62 +2,66 @@
 require_relative './basic_decorator'
 
 # NOTES
-# Appsignal interface to send our usual actions 
+# Appsignal interface to send our usual actions
 # BasicDecorator adds a transparent interface on top of the appsignal interface
 #
-# Usage: EventTracer.register :appsignal, EventTracer::AppsignalLogger.new(Appsignal)
-#        appsignal_logger.info appsignal: { increment_counter: { counter_1: 1, counter_2: 2 }, set_gauge: { gauge_1: 1 } }
-#        appsignal_logger.info appsignal: { set_gauge: { gauge_1: { value: 1, tags: { region: 'eu' } } } }
+# Usage: EventTracer.register :appsignal,
+#          EventTracer::AppsignalLogger.new(Appsignal, allowed_tags: ['tag_1', 'tag_2'])
+#
+#        appsignal_logger.info metrics: [:counter_1, :counter_2]
+#        appsignal_logger.info metrics: { counter_1: { type: :counter, value: 1 }, gauce_2: { type: :gauce, value: 10 } }
 module EventTracer
   class AppsignalLogger < BasicDecorator
 
-    SUPPORTED_METRICS ||= %i(increment_counter add_distribution_value set_gauge)
+    SUPPORTED_METRIC_TYPES = {
+      counter: :increment_counter,
+      distribution: :add_distribution_value,
+      gauge: :set_gauge
+    }
+    DEFAULT_METRIC_TYPE = :increment_counter
+    DEFAULT_COUNTER = 1
+
+    def initialize(decoratee, allowed_tags: [])
+      super(decoratee)
+      @allowed_tags = allowed_tags
+    end
 
     LOG_TYPES.each do |log_type|
       define_method log_type do |**args|
-        return LogResult.new(false, "Invalid appsignal config") unless args[:appsignal] && args[:appsignal].is_a?(Hash)
+        metrics = args[:metrics]
+
+        return fail_result('Invalid appsignal config') unless valid_args?(metrics)
+        return success_result if metrics.empty?
+
+        tags = allowed_tags.empty? ? {} : args.slice(*allowed_tags)
+
+        case metrics
+        when Array
+          metrics.each do |metric|
+            appsignal.public_send(DEFAULT_METRIC_TYPE, metric, DEFAULT_COUNTER, tags)
+          end
+        when Hash
+          metrics.each do |metric_name, metric_payload|
+            payload_type = metric_payload[:type]
+
+            unless payload_type.is_a?(String) || payload_type.is_a?(Symbol)
+              return fail_result("Appsignal metric #{payload_type} invalid")
+            end
```

Comment:
Could we skip this and let the error be raised instead?

Reply:
if we don't check against this, it will raise `NoMethodError: undefined method 'to_sym' for x:Class`, not sure is that a valid error that is good to show to the public?

Or we do a 
```ruby
    metric_type = SUPPORTED_METRIC_TYPES[payload_type.to_sym]
rescue NoMethodError => e
   return fail_result(xmessage) # or we refine the error message here?
```

am I understand correctly?

Reply:
> Not sure is that a valid error that is good to show to the public?

The error won't be shown to the public.

> if we don't check against this, it will raise NoMethodError: undefined method 'to_sym' for x:Class,

If metric cannot be converted to symbol, I think it's a bug from the developer side, and we should raise an error so that they can fix it early. Returning a failure result is only helpful when we don't see log in staging/prod and need investigation to see what is the issue.